File: 2. Structural Design Patterns/Composite/c++/CompositeTemplate.cpp
#include <iostream>
#include <vector>
using namespace std;

// Component
class Component {
public:
    virtual void operation() = 0;
    virtual void add(Component* component) {}
    virtual void remove(Component* component) {}
    virtual ~Component() {}
};

// Leaf
class Leaf : public Component {
public:
    void operation() override {
        cout << "Leaf operation\n";
    }
};

// Composite
class Composite : public Component {
private:
    vector<Component*> children;
public:
    void operation() override {
        cout << "Composite operation\n";
        for (auto child : children)
            child->operation();
    }

    void add(Component* component) override {
        children.push_back(component);
    }

    void remove(Component* component) override {
        children.erase(remove(children.begin(), children.end(), component), children.end());
    }

    ~Composite() {
        for (auto child : children)
            delete child;
    }
};

File: 2. Structural Design Patterns/Composite/Python/CompositeTemplate.py
# Composite Pattern – Python Template

from abc import ABC, abstractmethod

class Component(ABC):
    @abstractmethod
    def operation(self):
        pass

class Leaf(Component):
    def operation(self):
        return "Leaf operation"

class Composite(Component):
    def __init__(self):
        self.children = []
    
    def add(self, component):
        self.children.append(component)
    
    def remove(self, component):
        self.children.remove(component)
    
    def operation(self):
        results = []
        for child in self.children:
            results.append(child.operation())
        return f"Composite({', '.join(results)})"

def main():
    leaf1 = Leaf()
    leaf2 = Leaf()
    composite = Composite()
    composite.add(leaf1)
    composite.add(leaf2)
    print(composite.operation())

if __name__ == "__main__":
    main()
